[RFC] Improved auto preview

shaggyrogers opened the pull request:
This adds 2 new options:

* g:mundo_auto_preview_delay

When this option is set to a positive integer and auto preview is enabled, previews will only be rendered automatically after the cursor is stationary for g:mundo_auto_preview_delay milliseconds. Defaults to 0.

* g:mundo_inline_undo_delay

When this option is set, auto preview is enabled and g:mundo_auto_preview_delay is a positive integer, inline diffs will only be rendered after the cursor is stationary for g:mundo_auto_preview_delay milliseconds. Disabled by default.

dsummersl commented:
I'm sorry for the delay reviewing this, I'll get to it soon!

dsummersl commented on autoload/mundo.vim:
```diff
@@ -409,12 +411,59 @@ function! s:MundoRefresh()"{{{
-    autocmd CursorHold * call s:MundoRefresh()
-    autocmd CursorMoved * call s:MundoRefresh()
+    autocmd CursorMoved __Mundo__ call s:MundoUpdateCursor()
+    autocmd BufLeave __Mundo__ call s:MundoStopCursorTimer()
```
Restricting this autocmd to the __Mundo__ window means that users that leave the __Mundo__ window open and make changes, will not see their changes updated.

dsummersl replied:
(I think you'll want to use *)

shaggyrogers replied:
My bad, I should have picked this up while testing. Should be an easy fix.

dsummersl commented on autoload/mundo.vim:
```diff
@@ -409,12 +411,59 @@ function! s:MundoRefresh()"{{{
+endfunction"}}}
+
+function! s:MundoRenderPreviewDelayed(...)"{{{
+    if s:auto_preview_line == line('.')
```
There should be some guards here, similar to the guards in the MundoRefresh() method. Two bugs I'm seeing while testing now are:
 * if you call `:MundoShow`, and then move back to your main editor `control-wl`, the cursor focus will go back to to the __mundo__ window when the timer fires. Focus should remain where it was (see the MundoRefresh method to see how that happens currently).
 * if you call `:MundoShow` and then call `:MundoHide`, the __mundo_preview__ window will reopen when the timer fires.

dsummersl commented on doc/mundo.txt:
```diff
@@ -20,11 +20,13 @@ CONTENTS                                                      *Mundo-contents*
+        3.14 mundo_mirror_graph ........ |mundo_mirror_graph|
+        3.15 mundo_inline_undo ......... |mundo_inline_undo|
+        3.16 mundo_inline_undo_delay ... |mundo_inline_undo_delay|
+        3.17 mundo_return_on_revert .... |mundo_return_on_revert|
```
I can't think of a reason a user would want to use two different delay values - how about just having one `mundo_render_delay` (or whatever) that is used for both options?

shaggyrogers replied:
There's just the one delay value, the 'mundo_inline_undo_delay' option determines whether inline diff generation is also delayed. The name is misleading however, I'll change it to something a little more clearer, maybe 'mundo_enable_inline_delay'.

simnalamburt commented:
Thanks for the detailed review, dsummersl!

shaggyrogers commented:
I agree with everything you've said, thanks for taking the time to look over my PR. I'll implement the fixes you suggested and change the option name to something a little less misleading.

dsummersl commented:
Sorry for the late reply, sounds great - happy to chime in again when you're ready.

dsummersl commented:
@shaggyrogers hows things with this PR? If you don't mind I can help do some work on this with my somewhat limited free time. I think this'd be a great addition to mundo!

shaggyrogers commented:
Apologies, I've been busy lately and haven't got around to doing it. All that's left is to fix the issues @dsummersl pointed out and test everything again more thoroughly. I'll have time to finish it off on the weekend, but you are welcome to work on it if you want.

shaggyrogers commented:
Fixed the issues @dsummersl mentioned. There are a couple of issues with graph updates when editing with mundo open that I still need to finish fixing.

shaggyrogers commented:
I fixed the graph update issues I mentioned, performed some more thorough testing and cleaned up the new code a bit. 

I noticed a couple of issues, one that occurs with very large undo trees and appears to be unrelated to the PR:

<details><summary>Traceback</summary>
<p>

```Error detected while processing function <SNR>215_MundoRefreshDelayed[12]..mundo#MundoRenderGraph[5]..<SNR>215_MundoPython [2]..provider#python3#Call:
line   18:
Traceback (most recent call last):
File "<string>", line 1, in <module>
File "/home/.../vim-mundo/autoload/mundo.py", line 107, in MundoRenderGraph
    nodesData
File "/home/.../vim-mundo/autoload/mundo/graphlog.py", line 174, in generate
    current = nodesData.current()
File "/home/.../vim-mundo/autoload/mundo/node.py", line 91, in current
    nodes, nmap = self.make_nodes()
File "/home/.../vim-mundo/autoload/mundo/node.py", line 68, in make_nodes
    ut = util.vim().eval('undotree()')
File "/home/.../.local/lib/python3.6/site-packages/neovim/plugin/script_host.py", line 187, in eval
    obj = self.request("vim_eval", expr)
File "/home/.../.local/lib/python3.6/site-packages/neovim/api/nvim.py", line 141, in request
    return walk(self._from_nvim, res, decode=decode)
File "/home/.../.local/lib/python3.6/site-packages/neovim/api/common.py", line 176, in walk
    obj.items()
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 175, in <genexpr>
    return dict((walk(fn, k, *args), walk(fn, v, *args)) for k, v in
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 173, in walk
    return list(walk(fn, o, *args) for o in obj)
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 173, in <genexpr>
    return list(walk(fn, o, *args) for o in obj)
File "/home/.../.local/lib/python3.6/site-packages/neovim/api/common.py", line 176, in walk
    obj.items()
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 175, in <genexpr>
    return dict((walk(fn, k, *args), walk(fn, v, *args)) for k, v in
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 173, in walk
    return list(walk(fn, o, *args) for o in obj)
File  "/home/michael/.local/lib/python3.6/site-packages/neovim/api/common.py", line 173, in <genexpr>
    return list(walk(fn, o, *args) for o in obj)
...
File "/home/.../.local/lib/python3.6/site-packages/neovim/api/common.py", line 176, in walk
    obj.items()
Press ENTER or type command to continue
Error detected while processing function <SNR>215_MundoRefreshDelayed[12]..mundo#MundoRenderGraph[5]..<SNR>215_MundoPython[2]..provider#python3#Call:
line   18:
RecursionError: maximum recursion depth exceeded
Press ENTER or type command to continue
```
</p>
</details>
</br>

The other is that the new inline diffs don't render when editing, until moving the cursor into the graph window. This occurs after opening mundo in one tab, then opening another tab and opening mundo in it as well. I'm not sure what causes this, but I noticed that the 'wrap' option is reset in the mundo windows when this happens, so it may have something to do with the use of the BufNewFile autocommand to configure the graph window.

I'd appreciate if you could have another look and see if I've missed anything.

dsummersl commented:
So far it looks pretty good, but I still need to dig into the two problems you mentioned (I've been able to reproduce the 'insert space' problem you mentioned - not quite sure why `CursorHold` wouldn't take care of that).

dsummersl commented:
Since the 'insert space' problem exists outside your PR, I've opened a separate issue for that (#67).

dsummersl commented on doc/mundo.txt:
```diff
@@ -241,23 +253,32 @@ for a faster playback.
-3.15 g:mundo_return_on_revert                         *mundo_return_on_revert*
+3.17 g:mundo_return_on_revert                         *mundo_return_on_revert*
 
 Set this to 0 to keep focus in the Mundo window after a revert.
```
please add a summary of this feature the the changelog at the bottom of this file.

shaggyrogers replied:
Done. I'm assuming the project roughly follows [semantic versioning](https://semver.org/)? If so, it might be worth making a note of this in the documentation and/or README.

dsummersl replied:
Yeah, it *roughly* follows semantic versioning - or at least I'd like us to try to do that. If you have it in you to add a note about that, please do!

dsummersl commented on autoload/mundo.vim:
```diff
@@ -385,28 +388,88 @@ function! mundo#MundoHide()"{{{
-  let mundoPreWin = bufwinnr('__Mundo_Preview__')
-  let currentWin  = bufwinnr('%')
+    " Delayed refresh
+    if !get(g:, 'mundo_enable_inline_delay', 0) ||
```
Since `mundo_enable_inline_delay` isn't actually used for any `timer_start` call, it is effectively only a boolean property that determines whether `mundo#MundoRenderGraph()` is called in `MundoRefresh()` or `MundoRefreshDelayed()`. I think you should remove the setting altogether.

dsummersl commented on doc/mundo.txt:
```diff
@@ -241,23 +253,32 @@ for a faster playback.
 
+-----------------------------------------------------------------------------
+3.16 g:mundo_enable_inline_delay                   *mundo_enable_inline_delay*
+
```
As mentioned in comments above, this setting effectively does nothing. I don't see a benefit to having two different _delay settings either - I suggest we remove it.

dsummersl commented on doc/mundo.txt:
```diff
@@ -223,7 +225,17 @@ be useful on large files and undo trees to speed up Mundo.
+
+This is the delay in milliseconds before a preview diff is automatically 
+rendered. The delay is reset whenever the cursor moves. Use this to speed up 
+undo tree traversal when automatic previews are enabled.
```
Rather than this being a setting AND a delay value it should just be a delay value (say 100 or 200?)...I can't think of a reason we'd want to let the user turn this feature off...if they don't like it they can set a very small value, effectively turning it off.

shaggyrogers replied:
Agreed. 

After playing around with this option, ~250ms seems like a good choice, the preview delay is short but not so short that a user might scroll a little too slowly and trigger a preview, which may cause an annoying delay with larger files.
@simnalamburt any thoughts on a default?

dsummersl replied:
Sounds great - I tried 100 and 200 and was happy with both.

dsummersl commented on autoload/mundo.vim:
```diff
@@ -385,28 +388,88 @@ function! mundo#MundoHide()"{{{
+    if !get(g:, 'mundo_enable_inline_delay', 0) ||
+                \ !get(g:, 'mundo_inline_undo')
+        call mundo#MundoRenderGraph()
+    endif
```
If you remove the `mundo_enable_inline_delay` setting, you can remove this entire if-block - then there is only one place that `mundo#MundoRenderGraph()` will need to be called (and it won't be this place).

dsummersl commented on autoload/mundo.py:
```diff
@@ -262,7 +262,8 @@ def MundoMove(direction,move_count=1,relative=True,write=False):
 
-    if vim.eval('g:mundo_auto_preview') == '1':
+    if (vim.eval('g:mundo_auto_preview') != '0' and
+            int(vim.eval("get(g:, 'mundo_auto_preview_delay', 0)")) <= 0):
```
(see comments below) If we go with `mundo_auto_preview_delay` as a non-zero value, there is no need for this check to see if it is set here.

dsummersl commented on autoload/mundo.vim:
```diff
@@ -385,28 +388,88 @@ function! mundo#MundoHide()"{{{
+        return s:MundoRestartRefreshTimer()
+    endif
+
+    if get(g:, 'mundo_enable_inline_delay', 0) && get(g:, 'mundo_inline_undo')
```
As mentioned in the comments above, I suggest removing the `mundo_enable_inline_delay` part of this if-block.

shaggyrogers commented:
Cheers, I'll hopefully get the changes done by tonight.

shaggyrogers commented:
Whoops...

shaggyrogers commented:
Done. I also forgot to mention another small change, Mundo will no longer open for windows that aren't modifiable, are a preview window or have a buffer type other than the empty string and 'acwrite'. A reason message is displayed that indicates why the potential target buffer is considered invalid.

Issue - with automatic previews off and delay on, manual previews will revert after the delay time.
Need to test more thoroughly & on vim
Seems to be crash issue with neovim?

dsummersl commented:
Resolve the conflicts, and I think this is ready to go!

shaggyrogers commented:
I fixed a few more issues and added the other changes to the changelog. I just need to resolve the conflicts and fix an issue with Mundo and different tabs that I introduced somewhere in the last commit and I'll be done.

shaggyrogers commented:
Here are the changes, from the changelog:

    * Adds g:mundo_auto_preview_delay setting.
    * Adds validation for target buffers.
    * Adds 'G' mapping to graph.
    * Fix a runtime error when entering embedded terminal windows in neovim.
    * Fix multiple issues associated with Mundo windows in different tabs.
    * Fix default option values overriding user-set values in certain cases.
    * Fix cursor initially being positioned above the graph.
    * Fix some mappings not working correctly when preceded with a count.
    * Fix some graph mappings moving the cursor to the target buffer.
    * Fix airline plugin ignoring custom status lines.
    * Fix Escape and Ctrl+c not being handled correctly in the search prompt.
    * Fix MundoGetTargetState() returning invalid results with some inline diffs.
    * Reduced delay associated with animated playback for neovim.
    * Code cleanup and refactoring.

There are a few minor bugfixes and improvements that I didn't think were significant enough to warrant listing in the changelog:

* Fixed message when opening Mundo
* Fixed 'cpoptions' not being restored on initialisation failure
* MundoClose() now closes multiple windows and windows open in other tabs
* MundoPlayTo() now restores folds to their prior state

Some of the cleanup/refactoring:

* Decoupled MundoRenderGraph() and MundoRenderPreview()
* Merged util._goto_window_for_buffer_name() into util._goto_window_for_buffer()
* Made MundoOpenGraph() and MundoOpenPreview() more concise
* Moved some generic functions from autoload/mundo.vim to util.vim
* Moved code to set option defaults to plugin/mundo.vim

This also should fix issues #62 and #67. 

I obviously got a little carried away, I'm happy to create separate PRs or revert some of the changes if necessary.

dsummersl commented:
Thanks @shaggyrogers  I've been trying it out and trying to review the changes - I'll keep looking this week.

dsummersl commented:
This has worked well for me over the past month - I think we should merge it!
